src/Codeception/Module/MarkupValidator.php

```diff
@@ -0,0 +1,184 @@
+<?php
+namespace Codeception\Module;
+
+use Exception;
+use stdClass;
+use Codeception\Lib\Interfaces\PageSourceViewer;
+use Codeception\Module;
+use GuzzleHttp\Client;
+
+/**
+ * A module which validates current page markup via the W3C Markup Validation Service.
+ * Requires either the `PhpBrowser` or the `WebDriver` module.
+ *
+ * Configuration options:
+ *  - pageSourceViewer
+ *  - ignoreWarnings
+ *  - ignoredErrors
+ */
+class MarkupValidator extends Module
+{
+    const W3C_MARKUP_VALIDATION_SERVICE_BASE_URI = 'https://validator.w3.org';
+
+    const W3C_MARKUP_VALIDATION_SERVICE_ENDPOINT = '/nu/';
+
+    /**
+     * Validates current page markup via the W3C Markup Validation Service.
+     *
+     * @param bool|null $ignoreWarnings Whether to ignore warnings or not.
+     * If `null`, module-wide value is used. If module-wide value is `null` too
+     * then warnings are not ignored by default.
+     */
+    public function validateMarkup($ignoreWarnings = null)
+    {
+        $markup = $this->getCurrentPageMarkup();
+        $validationData = $this->sendMarkupValidationRequest($markup);
+        foreach ($validationData->messages as $message) {
+            $this->processMarkupValidationMessage($message, $ignoreWarnings);
+        }
+    }
+
+    /**
+     * Returns current page markup.
+     *
+     * @return string Current page markup.
+     */
+    protected function getCurrentPageMarkup()
+    {
+        $pageViewerConfigKey = 'pageSourceViewer';
+        if (!isset($this->config[$pageViewerConfigKey])) {
+            throw new Exception(sprintf(
+                'Missing mandatory configuration parameter "%s".',
+                $pageViewerConfigKey
+            ));
+        }
+
+        $pageViewerName = $this->config[$pageViewerConfigKey];
+        if (!$this->hasModule($pageViewerName)) {
+            throw new Exception(sprintf(
+                '"%s" page viewer was not found.',
+                $pageViewerName
+            ));
+        }
+
+        /* @var $pageViewer PageSourceViewer */
+        $pageViewer = $this->getModule($pageViewerName);
```

Comment:
Markup validator will have to know implementation details of each dependent module to get page source from them which is not good.

I think it's better to extract common functionality to an interface. This will allow client code independent from implementation details of dependent modules. This will lower module coupling which is always good.